### PR TITLE
fix: use explicit api versions for agent and tailnet

### PR DESCRIPTION
**Pull request by spikecurtis**
Bumps the Tailnet and Agent API version 2.3, and creates some extra controls and machinery around these versions.

What happened is that we accidentally shipped two new API features without bumping the version.  `ScriptCompleted` on the Agent API in Coder v2.16 and `RefreshResumeToken` on the Tailnet API in Coder v2.15.

Since we can't easily retroactively bump the versions, we'll roll these changes into API version 2.3 along with the new WorkspaceUpdates RPC, which hasn't been released yet.  That means there is some ambiguity in Coder v2.15-v2.17 about exactly what methods are supported on the Tailnet and Agent APIs.  This isn't great, but hasn't caused us major issues because 

1. RefreshResumeToken is considered optional, and clients just log and move on if the RPC isn't supported. 
2. Agents basically never get started talking to a Coderd that is older than they are, since the agent binary is normally downloaded from Coderd at workspace start.

Still it's good to get things squared away in terms of versions for SDK users and possible edge cases around client and server versions.

To mitigate against this thing happening again, this PR also:

1. adds a CODEOWNERS for the API proto packages, so I'll review changes
2. defines interface types for different API versions, and has the agent explicitly use a specific version.  That way, if you add a new method, and try to use it in the agent without thinking explicitly about versions, it won't compile.

With the protocol controllers stuff, we've sort of already abstracted the Tailnet API such that the interface type strategy won't work, but I'll work on getting the Controller to be version aware, such that it can check the API version it's getting against the controllers it has -- in a later PR.

### CODE REVIEW

**Comment by spikecurtis** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/coder/coder/15508?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15508** <a href="https://app.graphite.dev/github/pr/coder/coder/15508?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#15507** <a href="https://app.graphite.dev/github/pr/coder/coder/15507?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#15506** <a href="https://app.graphite.dev/github/pr/coder/coder/15506?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @spikecurtis and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: agent/agent.go
Comment (johnstcn) [inline agent/agent.go:2026]: nit: The `a`, `b` variable naming is somewhat confusing here. I'd suggest renaming `b` to something more easily recognizable.

**Review by johnstcn — APPROVED** [review]
Some comments but nothing blocking from my side.

File: CODEOWNERS
Comment (spikecurtis) [reply CODEOWNERS:6]: Who should it be? 🤔

Comment (johnstcn) [reply CODEOWNERS:6]: Dealer's choice ;-)

File: agent/agent.go
Comment (mafredri) [inline agent/agent.go:2016]: What's the difference between this and `startTailnetAPI`? These look like identical functions to me. 🤔

Edit: Now I see it, it's what's passed to the func (aAPI/tAPI). Perhaps a comment would be good since the difference is hard to spot via ocular observation.

File: codersdk/agentsdk/agentsdk.go
Comment (mafredri) [inline codersdk/agentsdk/agentsdk.go:230]: Could some or most of these changes be automated via code-gen? How easy is it for someone unfamiliar with this domain to add a new API version without forgetting something in the process?

File: enterprise/wsproxy/wsproxysdk/wsproxysdk.go
Comment (mafredri) [inline enterprise/wsproxy/wsproxysdk/wsproxysdk.go:518]: Suggestion: We could move `"2.0"` to a const and the comment as well? This way it has a bit more exposure outside the package. Just thinking if it's a bit too hidden whilst making assumptions(?) about current business logic/requirements.

**Review by mafredri — APPROVED** [review]
Does this PR cover the case where proto changes are introduced without bumping the version? If it is I didn't quite catch how that results in a build failure. (I.e. is anything stopping me from appending a new method to `DRPCAgentClient23`?)

Other than that, looks good. If we can automate more of this process in the future that'd be good. 👍🏻

**Comment by spikecurtis** [comment]
> Does this PR cover the case where proto changes are introduced without bumping the version? If it is I didn't quite catch how that results in a build failure. (I.e. is anything stopping me from appending a new method to DRPCAgentClient23?)

My assumption is that we just edited `agent.proto` and `tailnet.proto`, then ran `make gen` and forgot/didn't know we needed to care about the version.  So, having a separate `DRPCAgentClient23` piped around the `agent` means that when you go to use your fancy new method, it isn't on the interface and your build will fail.  You _could_ just add it to the interface manually, but my hope is that the comment on that interface makes it a bit more obvious that this is a versioned interface and you shouldn't do that.  Before we were just passing around the `drpc.Conn`, and constructing a `DRPCAgentClient`, which is autogenerated from the `.proto`, so it was very easy not to realize the protocol is versioned.

File: codersdk/agentsdk/agentsdk.go
Comment (spikecurtis) [reply codersdk/agentsdk/agentsdk.go:230]: Hard to automate directly from the `.proto` because it requires knowing which methods were added in which version.

We could annotate the RPCs somehow, in comments, but I'm hesitant to do that.

1. As a matter of automation, I don't think the time spent to implement will save us time overall, considering this is something we change a few times per year.
2. What I'm doing here only covers API changes that add new _methods_. You could also imagine adding new fields to existing messages in a back-compatible way.
3. I don't want to entrench us too much into this particular paradigm by building a bunch of infrastructure around it. I'm not totally sure this is the right long-term way to manage the versions, so I want to keep it lightweight. We might change it depending on future changes to the API.

**Comment by mafredri** [comment]
@spikecurtis ok, thanks for the explanation. This got me thinking though, could we instead make the protocol version a type alias so you don't need to change all the code every time you're switching? I.e.

```go
type DRPCAgentClient = proto.DRPCAgentClient23
```

**Comment by Emyrk** [comment]
> @spikecurtis ok, thanks for the explanation. This got me thinking though, could we instead make the protocol version a type alias so you don't need to change all the code every time you're switching? I.e.
> 
> ```go
> type DRPCAgentClientLatest = DRPCAgentClient23
> 
> func (c *Client) ConnectRPCLatest(...) { return c.ConnectRPC23(...) }
> ```
> 
> Edit: Thinking about this some more, this would essentially automatically bump the protocol in the common case for all consumers when you change the alias. Maybe that's not what we want. I was hoping to avoid the yank renaming across the code-base but may be unavoidable without some rethinking of the API.

@mafredri I think if you did this, maybe we could do something with golden files and dump a JSON representation of the interface somewhere. If it changes, via a argument field change, new method, etc. The test would fail until you `update-golden-files` again.

Just an idea to "snapshot" what the API is from a JSON POV and dump it into a test that can detect changes.

@spikecurtis I know you mentioned it might not be worth spending a lot of time on this infrastructure. I'll defer to ya'll on it.

**Comment by spikecurtis** [comment]
> maybe we could do something with golden files and dump a JSON representation of the interface somewhere. If it changes, via a argument field change, new method, etc. The test would fail until you update-golden-files again.
>
> Just an idea to "snapshot" what the API is from a JSON POV and dump it into a test that can detect changes.

Look, if we wanted to snapshot the API definition, we could just make a copy of the `.proto` file. No need to involve JSON.

This isn't like a "golden file" where you're testing that the code gives some complex output that's easier to just compare wholesale rather than asserting individual facts about it.

We could stash a snapshot of the current version in testdata and directly compare, so that if you edited the proto without creating a new version, tests would fail. Personally, I think CODEOWNERS should help here, but if y'all really want another robot to complain to you about stuff, we can do it.

**Comment by Emyrk** [comment]
>  we could just make a copy of the `.proto` file. No need to involve JSON.

Oh yea :facepalm:. Nah, this is all fine with me :+1:

**Comment by mafredri** [comment]
Let's go with this for now, we can always revisit if the process is too cumbersome or has other shortcomings 👍🏻.

**Review by deansheather — APPROVED** [review]

**Comment by spikecurtis** [comment]
### Merge activity

* **Nov 15, 2:16 AM EST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/coder/coder/15508).

**Pull request merged**
